### PR TITLE
Peabody/cse matrixsymbol

### LINKED ISSUE #11991: cse() has strange behaviour for MatrixSymbol indexing
Example: 
```python
import sympy as sp
from pprint import pprint


def sub_in_matrixsymbols(exp, matrices):
    for matrix in matrices:
        for i in range(matrix.shape[0]):
            for j in range(matrix.shape[1]):
                name = "%s_%d_%d" % (matrix.name, i, j)
                sym = sp.symbols(name)
                exp = exp.subs(sym, matrix[i, j])
    return exp


def t44(name):
    return sp.Matrix(4, 4, lambda i, j: sp.symbols('%s_%d_%d' % (name, i, j)))


# Construct matrices of symbols that work with our
# expressions. (MatrixSymbols does not.)
a = t44("a")
b = t44("b")

# Set up expression. This is a just a simple example.
e = a * b

# Put in matrixsymbols. (Gives array-input in codegen.)
e2 = sub_in_matrixsymbols(e, [sp.MatrixSymbol("a", 4, 4), sp.MatrixSymbol("b", 4, 4)])
cse_subs, cse_reduced = sp.cse(e2)
pprint((cse_subs, cse_reduced))

# Codegen, etc..
print "\nccode:"
for sym, expr in cse_subs:
    constants, not_c, c_expr = sympy.printing.ccode(
        expr,
        human=False,
        assign_to=sympy.printing.ccode(sym),
    )
    assert not constants, constants
    assert not not_c, not_c
    print "%s\n" % c_expr

```

This gives the following output:

```
([(x0, a),
  (x1, x0[0, 0]),
  (x2, b),
  (x3, x2[0, 0]),
  (x4, x0[0, 1]),
  (x5, x2[1, 0]),
  (x6, x0[0, 2]),
  (x7, x2[2, 0]),
  (x8, x0[0, 3]),
  (x9, x2[3, 0]),
  (x10, x2[0, 1]),
  (x11, x2[1, 1]),
  (x12, x2[2, 1]),
  (x13, x2[3, 1]),
  (x14, x2[0, 2]),
  (x15, x2[1, 2]),
  (x16, x2[2, 2]),
  (x17, x2[3, 2]),
  (x18, x2[0, 3]),
  (x19, x2[1, 3]),
  (x20, x2[2, 3]),
  (x21, x2[3, 3]),
  (x22, x0[1, 0]),
  (x23, x0[1, 1]),
  (x24, x0[1, 2]),
  (x25, x0[1, 3]),
  (x26, x0[2, 0]),
  (x27, x0[2, 1]),
  (x28, x0[2, 2]),
  (x29, x0[2, 3]),
  (x30, x0[3, 0]),
  (x31, x0[3, 1]),
  (x32, x0[3, 2]),
  (x33, x0[3, 3])],
 [Matrix([
[    x1*x3 + x4*x5 + x6*x7 + x8*x9,     x1*x10 + x11*x4 + x12*x6 + x13*x8,     x1*x14 + x15*x4 + x16*x6 + x17*x8,     x1*x18 + x19*x4 + x20*x6 + x21*x8],
[x22*x3 + x23*x5 + x24*x7 + x25*x9, x10*x22 + x11*x23 + x12*x24 + x13*x25, x14*x22 + x15*x23 + x16*x24 + x17*x25, x18*x22 + x19*x23 + x20*x24 + x21*x25],
[x26*x3 + x27*x5 + x28*x7 + x29*x9, x10*x26 + x11*x27 + x12*x28 + x13*x29, x14*x26 + x15*x27 + x16*x28 + x17*x29, x18*x26 + x19*x27 + x20*x28 + x21*x29],
[x3*x30 + x31*x5 + x32*x7 + x33*x9, x10*x30 + x11*x31 + x12*x32 + x13*x33, x14*x30 + x15*x31 + x16*x32 + x17*x33, x18*x30 + x19*x31 + x20*x32 + x21*x33]])])

ccode:
x0[0] = a[0];
x0[1] = a[1];
x0[2] = a[2];
x0[3] = a[3];
x0[4] = a[4];
x0[5] = a[5];
x0[6] = a[6];
x0[7] = a[7];
x0[8] = a[8];
x0[9] = a[9];
x0[10] = a[10];
x0[11] = a[11];
x0[12] = a[12];
x0[13] = a[13];
x0[14] = a[14];
x0[15] = a[15];
x1 = x0[0];
x2[0] = b[0];
x2[1] = b[1];
x2[2] = b[2];
x2[3] = b[3];
x2[4] = b[4];
x2[5] = b[5];
x2[6] = b[6];
x2[7] = b[7];
x2[8] = b[8];
x2[9] = b[9];
x2[10] = b[10];
x2[11] = b[11];
x2[12] = b[12];
x2[13] = b[13];
x2[14] = b[14];
x2[15] = b[15];
x3 = x2[0];
x4 = x0[1];
x5 = x2[4];
x6 = x0[2];
x7 = x2[8];
x8 = x0[3];
x9 = x2[12];
x10 = x2[1];
x11 = x2[5];
x12 = x2[9];
x13 = x2[13];
x14 = x2[2];
x15 = x2[6];
x16 = x2[10];
x17 = x2[14];
x18 = x2[3];
x19 = x2[7];
x20 = x2[11];
x21 = x2[15];
x22 = x0[4];
x23 = x0[5];
x24 = x0[6];
x25 = x0[7];
x26 = x0[8];
x27 = x0[9];
x28 = x0[10];
x29 = x0[11];
x30 = x0[12];
x31 = x0[13];
x32 = x0[14];
x33 = x0[15];
```

`x0` and `x2` are just copies of the matrices `a` and `b`, respectively.

### PR DESCRIPTION
**Pull request by zouhairm**
Treat matrix symbols and subelements as atomic instead of redefining them as individual variables.

fixes #11991 by reviving PR #13185 

Thanks @peabody124 for original fix, hopefully with release notes we can merge this @czgdp1807 

(couldn't push directly to that PR since I'm not a maintainer)

* simplify
  * CSE now treats matrix symbols and subelements as atomic in cse

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
Can you explain more what this change does?

This PR would need tests before it could be merged. (Tests can help explain what it does.)

**Comment by zouhairm** [comment]
Hi @oscarbenjamin.

This fixes #11991 by reviving a staled PR #13185 that the original author didn't end up merging.

I stumbled on #11991 after having the same issue of cse (used within code generation) which was creating temporary variables to store a copy of a matrix as opposed to directly indexing into the matrix.

The original author did provide a test, but this was done before code generation actually supported matrixsymbols. I think the [test](https://github.com/sympy/sympy/pull/22840/files#diff-5ebea3a9a6fb60b1a8487ae2b74a60d92751c7464eda8eafdb66bf8650686041) I updated probably best illustrates what this does

File: sympy/simplify/cse_main.py [hunk 1]
```diff
@@ -586,7 +587,7 @@ def _find_repeated(expr):
         if isinstance(expr, RootOf):
             return
 
-        if isinstance(expr, Basic) and (expr.is_Atom or expr.is_Order):
+        if isinstance(expr, Basic) and (expr.is_Atom or expr.is_Order or isinstance(expr, MatrixSymbol) or isinstance(expr, MatrixElement)):
```
Comment (smichr) [inline sympy/simplify/cse_main.py:590]:
```suggestion
        if isinstance(expr, Basic) and (
                expr.is_Atom or
                expr.is_Order or
                isinstance(expr, (MatrixSymbol, MatrixElement)):
```

File: sympy/simplify/cse_main.py [hunk 2]
```diff
@@ -586,7 +587,10 @@ def _find_repeated(expr):
         if isinstance(expr, RootOf):
             return
 
-        if isinstance(expr, Basic) and (expr.is_Atom or expr.is_Order):
+        if isinstance(expr, Basic) and (
+                expr.is_Atom or
+                expr.is_Order or
+                isinstance(expr, (MatrixSymbol, MatrixElement)):
```
Comment (smichr) [inline sympy/simplify/cse_main.py:593]:
```suggestion
                isinstance(expr, (MatrixSymbol, MatrixElement))):
```

**Comment by asmeurer** [comment]
This did not address the comments on the original PR. See https://github.com/sympy/sympy/pull/13185#issuecomment-1011479644. This is a regression and we should either revert this or make a better fix. With this change, no common subexpressions inside of matrix indices will be cse'd.

Incidentally, I don't really see what's wrong with cse treating something like `A[0, 0]` as a common subexpression as in the original issue (https://github.com/sympy/sympy/issues/11991). This does actually correspond to a computation (e.g., some pointer arithmetic, or in NumPy, an indexing expression that creates a view), and so can be worth computing only once.

**Comment by zouhairm** [comment]
@asmeurer  - apologies I had understood from [this comment](https://github.com/sympy/sympy/pull/13185#issuecomment-324776745) that index behavior of cse was already 'idiosyncratic' and therefore thought that it wasn't a factor with this PR.

For example, this is the old behavior:
```
>>> sp.cse(A[0, i + j] + A[0, i] + B[0, 2*(i + j)])
([(x0, A)], [B[0, 2*i + 2*j] + x0[0, i] + x0[0, i + j]])
```
The issue isn't creating A[0,0] as a common subexpression, it is rather the unnecessary creating of `x0=A`, which in code generation leads to an entire matrix copy.

The tradeoff with that in this merge is that instead of that expensive matrix copy, now indices are not being cse'd. IMO that's ok, but I see how the ideal behavior would be to have both behaviors. Perhaps we can open an issue for that and I can try to fix it in a different PR?

**Comment by asmeurer** [comment]
I agree that (x0, A) isn't good. The changes here set both `MatrixSymbol` and `MatrixElement` as atoms, but it should just be `MatrixSymbol`. That would prevent that case but still allow cse to process indices inside of a `MatrixElement`.

**Pull request merged**